adding labeling webhooks to ML for continuous delivery

babineaum opened the pull request:
not sure this is all we need, going to need a double check

ferrants commented:
fork/pull already works for all repos, this is for the continuous delivery workflow and looks fine

ferrants commented on map_configs/dx_map_config.yml:
```diff
@@ -84,7 +84,7 @@ pr:
 issue:
     - match:
-        repo: "(dxng|user-interface)"
+        repo: "(dxng|user-interface|machine-learning|machine-learning-*|dss-*)"
```
it uses regex matching, not glob matching, the stars need to be like this: `machine-learning-.*|dss-.*`

babineaum replied:
thanks

babineaum commented:
@ferrants question, will the "machine-learning.*" catch both the "machine-learning" and "machine-learning-monitor" and "machine-learning-spark". Same question for ".*data-slicing-service" should catch "data-slicing-service" as well as "apiary-data-slicing-service"???

ferrants commented:
no, it won't. it's a regular expression. "machine-learning." would match "machine-learning" followed by any one character, so "machine-learningX" would match, but "machine-learning" would not. "machine-learning.*"  would match "machine-learning" followed by any number (including 0) of any character, so ""machine-learning" and "machine-learningXXXX"

babineaum commented:
@ferrants thanks! yeah that makes sense, wasnt thinking about the .* or . meaning there has to be something there in order for it to match.

ferrants commented on map_configs/dx_map_config.yml:
```diff
@@ -84,7 +84,7 @@ pr:
 issue:
     - match:
-        repo: "(dxng|user-interface)"
+        repo: "(dxng|user-interface|machine-learning.*|machine-learning|dss-.*|auto-cruise-control|optimization.*|optimization|data-slicing-service|.*data-slicing-service|learning-config)"
```
these are redundant:
- `optimization.*|optimization`
- `machine-learning.*|machine-learning`
- `data-slicing-service|.*data-slicing-service`

mbomio replied:
we have some others, not sure if you want them or not:
`dss-common-java`
`dss-common-python`
`hive-functions`
`qubole-utilities`
`spend-predictor`
`dss-blacklisting`
`dss-blacklist-updater`

babineaum commented:
i didnt realize what i wrote was . instead of .*   

I meant to be asking about .*

should be all set now.

ferrants commented:
looks good, can you squash to one commit and then :shipit: if the teams are aware of the flow and ready to hop on board

babineaum commented:
ok one more look at this. added extra tests to make sure the full list of added repos made it through
and also added a negative test to make sure that one of the non-CD repos couldn't trigger.

dx-pbuckley commented:
:+1: I like how you split sampling the different repos over the different tests - time to try the new squash-n-shipit button? :shipit:

ferrants commented:
i'm not a fan of us having all these test cases for our specific configuration. We should have test cases to make sure that github-webhooks works and we should trust those test cases well enough that we can have our own configuration

babineaum commented:
How might you have handled adding tests for these additional changes to the map_config?

I think since the map config is very customized at this point it's hard not to have customized tests.

ferrants commented:
I would test the application, not our configuration for the application.

cchristous commented:
the DX specific configuration should live in a separate repo and be tested separately

ferrants commented:
So, i wouldn't have added tests for these changes, except maybe the negative case.

babineaum commented:
Ok so moving the configuration into another repo is possible. but for the time being not having a test for some of the newly added repos would not leave me feeling like I did due diligence in making sure that this would not break.

babineaum commented:
We can look into separating out the config and creating tests against that config in another repo

ferrants commented:
Tests really only need to be added for new or untested functionality of the application. We shouldn't test our specific configuration, we should trust that the spec is defined and works correctly

dx-pbuckley commented:
@ferrants I think where we have these kinds of long regexes in the map/config we want more than just a second set of eyes making sure there aren't any errors in there and that we are matching what we expect to match. I agree that it isn't a "unit test" or a "test of github-webhooks, the application" to test our configuration, but we should be testing our configuration when it contains "code" or code-like things (like a regex) that we depend on precisely matching/working or we have fix and then deploy the app again

dx-pbuckley commented:
also I should add I'm in favor of having our config live outside this repo, along with all our test data for testing the config - we could sanitize it less and it would def make for less churn/surface-area in what is an open source repo

ferrants commented:
Also, I'm not saying we should revert this and switch it all up right now. For the future, we don't need to add test cases here for every configuration change. We should eventually move out map_config file to dxansible and deploy it with the application. The application should have its own configuration for test. If we need to add test cases for how regular expressions work, we can add that.

With the config in dxansible, deploying a configuration change would not require rebuilding github-webhooks. This would make deployments very quick, so reverting a change in trivial.
